### PR TITLE
[SPARK-19313][ML][MLLIB] GaussianMixture should limit the number of features

### PR DESCRIPTION
**Pull request by sethah**
## What changes were proposed in this pull request?

The following test will fail on current master

````scala
test("gmm fails on high dimensional data") {
    val ctx = spark.sqlContext
    import ctx.implicits._
    val df = Seq(
      Vectors.sparse(GaussianMixture.MAX_NUM_FEATURES + 1, Array(0, 4), Array(3.0, 8.0)),
      Vectors.sparse(GaussianMixture.MAX_NUM_FEATURES + 1, Array(1, 5), Array(4.0, 9.0)))
      .map(Tuple1.apply).toDF("features")
    val gm = new GaussianMixture()
    intercept[IllegalArgumentException] {
      gm.fit(df)
    }
  }
````

Instead, you'll get an `ArrayIndexOutOfBoundsException` or something similar for MLlib. That's because the covariance matrix allocates an array of `numFeatures * numFeatures`, and in this case we get integer overflow. While there is currently a warning that the algorithm does not perform well for high number of features, we should perform an appropriate check to communicate this limitation to users.

This patch adds a `require(numFeatures < GaussianMixture.MAX_NUM_FEATURES)` check to ML and MLlib algorithms. For the feature limitation, we can limit it such that we do not get numerical overflow to something like `math.sqrt(Integer.MaxValue).toInt` (about 46k) which eliminates the cryptic error. However in, for example WLS, we need to collect an array on the order of `numFeatures * numFeatures` to the driver and we therefore limit to 4096 features. We may want to keep that convention here for consistency.

## How was this patch tested?
Unit tests in ML and MLlib.

### CODE REVIEW

**Comment by sethah** [comment]
ping @yanboliang

**Comment by yanboliang** [comment]
Will review it tomorrow. Thanks.

File: mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala [hunk 1]
```diff
@@ -486,6 +491,9 @@ class GaussianMixture @Since("2.0.0") (
 @Since("2.0.0")
 object GaussianMixture extends DefaultParamsReadable[GaussianMixture] {
 
+  /** Limit number of features such that numFeatures^2^ < Integer.MaxValue */
+  private[clustering] val MAX_NUM_FEATURES = 46000
```
Comment (imatiach-msft) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala]: shouldn't this be in upper camel case according to scala style?

http://docs.scala-lang.org/style/naming-conventions.html

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala [hunk 2]
```diff
@@ -272,6 +277,10 @@ class GaussianMixture private (
 }
 
 private[clustering] object GaussianMixture {
+
+  /** Limit number of features such that numFeatures^2^ < Integer.MaxValue */
+  private[clustering] val MAX_NUM_FEATURES = 46000
```
Comment (imatiach-msft) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: it looks like the constant can be shared between the two clustering algorithms - I would recommend using the mllib one for now.

File: mllib/src/test/scala/org/apache/spark/ml/clustering/GaussianMixtureSuite.scala [hunk 3]
```diff
@@ -53,6 +53,19 @@ class GaussianMixtureSuite extends SparkFunSuite with MLlibTestSparkContext
     rDataset = rData.map(FeatureData).toDF()
   }
 
+  test("gmm fails on high dimensional data") {
+    val ctx = spark.sqlContext
+    import ctx.implicits._
```
Comment (imatiach-msft) [inline mllib/src/test/scala/org/apache/spark/ml/clustering/GaussianMixtureSuite.scala]: is there a way to remove this import?  I'm not sure why you need it.

**Comment by imatiach-msft** [comment]
I left few minor comments, looks good to me!

**Review by imatiach-msft — APPROVED** [review]

File: mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala (hunk 1 above)
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala]: This is like a `private static final` field in Java, and when used for constants, CONSTANT_CASE is normal.

**Review by srowen — APPROVED** [review]

File: mllib/src/test/scala/org/apache/spark/ml/clustering/GaussianMixtureSuite.scala (hunk 3 above)
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/clustering/GaussianMixtureSuite.scala]: Removed.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala (hunk 2 above)
Comment (sethah) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: I can see benefits either way, but I think leaving ML GMM to be completely independent of MLlib is slightly preferable.

Comment (imatiach-msft) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: ok

**Comment by imatiach-msft** [comment]
LGTM, nice work!  Who has the permissions to push the changes?

**Comment by sethah** [comment]
@imatiach-msft Spark committers must push the changes. As long as at least one committer is aware of the changes there is probably nothing left to do.

**Comment by sethah** [comment]
Thanks for the review @srowen and @imatiach-msft!

File: mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala (hunk 1 above)
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala]: In https://github.com/apache/spark/pull/15413, the symmetry of covariance matrix is taken into account and only the upper triangular part is store. So this number seems to be 65535? (`math.sqrt(Int.MaxValue.toDouble * 2)`)

Comment (sethah) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala]: We have to unpack the covariance matrix to a full covariance matrix before returning the model.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala (hunk 2 above)
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: ultimately *long-term* the plan is to deprecate `mllib` so keeping it separate is preferable

File: mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala (hunk 1 above)
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala]: Is floor(sqrt(2^31-1)) = 46340 more accurate? or is there overhead that prevents this from being achievable? I know it's a corner case, but if 46000 is a number that's just "about" the real max, let's just use the real max.

Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala]: +1 @srowen It's better to use the real max.

File: mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala [hunk 4]
```diff
@@ -486,6 +491,9 @@ class GaussianMixture @Since("2.0.0") (
 @Since("2.0.0")
 object GaussianMixture extends DefaultParamsReadable[GaussianMixture] {
 
+  /** Limit number of features such that numFeatures^2^ < Integer.MaxValue */
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala]: Nit: ```Integer.MaxValue``` is not a standard convention, it should be ```Int.MaxValue``` in Scala or ```Integer.MAX_VALUE``` in Java.

**Review by yanboliang — COMMENTED** [review]
LGTM except for very minor comments.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala [hunk 5]
```diff
@@ -272,6 +277,10 @@ class GaussianMixture private (
 }
 
 private[clustering] object GaussianMixture {
+
+  /** Limit number of features such that numFeatures^2^ < Int.MaxValue */
+  private[clustering] val MAX_NUM_FEATURES = math.sqrt(Int.MaxValue).toInt
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: The number is not equal to that used in `computeCovariance()` in `mllib.linalg.distributed.RowMatrix`.
https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/RowMatrix.scala#L327
Do the limits in `mllib.linalg.distributed.RowMatrix` need to be updated to this one?

**Comment by zhengruifeng** [comment]
BTW, it maybe nice to add a `SymmetricMatrix` class, for symmetric matrice are widely used in computation of covariance/concurrence/etc

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala (hunk 5 above)
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: I believe the limiting factor here is that we can't have an array of elements somewhere that has more than 2^31 - 1 elements. For a dense representation of a normal n x n matrix, that limits n to 46340. Here, however, the matrix is a symmetric Gramian matrix that needs n(n+1)/2 elements of storage, so 65535 works.

**Comment by yanboliang** [comment]
Merged into master. Thanks for all.

**Pull request closed without merging**